fix(atc): prevent partial redaction of overlapping secrets

analytically opened the pull request:
Secrets redaction had two bugs:

1. When one secret was a substring of another (e.g., "pass" and
   "password"), map iteration order determined replacement order.
   If "pass" replaced first, "password" became "((redacted))word".

2. Secrets containing substrings of "((redacted))" (e.g., "redact")
   corrupted previously inserted markers, producing mangled output
   like "((((redacted))ed))".

Fix: use a bitmask to track redacted byte positions in the original
string. Consecutive masked regions collapse to a single "((redacted))"
marker in one final output pass.

Also use utf8.RuneCountInString for the single-char check to correctly
handle multi-byte characters.

taylorsilva commented on atc/engine/build_step_delegate_test.go:
```diff
@@ -1050,6 +1050,56 @@ var _ = Describe("BuildStepDelegate", func() {
+					Expect(payload).To(Equal("the ((redacted)) value"))
+					Expect(payload).ToNot(ContainSubstring("(((("))
+				})
+			})
```
Okay here's another idea. Instead of `((redacted))` being the string we replace things with, we use `((***))` or `***` instead? It reduces the surface area of the edge case from "all substrings of redacted" to three asterisks. A lone secret having a value of `*`, `**`, or `***` feels _incredibly_ unlikely.
Wdyt?

taylorsilva replied:
I am very against doing a string replacement of `\x00` because it makes things less predictable for users. As-is, users can understand what happens if they see `((((redacted))ed))`. They can change the value of the secret to avoid this issue.

With this PR, if any of the log output contains `\x00` a users log output would become a complete mess of `((redacted))`'s and it would be out of the users control, unlike in the current situation.

analytically replied:
Ok I'll change the marker. Make it more strange.

taylorsilva replied:
Was curious how we even settled on `((redacted))`. The history:
- Started as `[***redacted**]` in the original PR https://github.com/concourse/concourse/pull/4311
- Became `((redacted))` in https://github.com/concourse/concourse/pull/4526 because Marie Kondo vibes ✨

analytically replied:
Yes, that's the whole point. X00 doesn't transmit, doesn't show. So use it as marker in a byte seq. But I've moved on, it's a bitmask now, better in all cases.

analytically replied:
Bitmask now.

analytically committed:
fix(atc): prevent partial redaction of substring secrets

Secrets redaction had two bugs:

1. When one secret was a substring of another (e.g., "pass" and
   "password"), map iteration order determined replacement order.
   If "pass" replaced first, "password" became "((redacted))word".

2. Secrets containing substrings of "((redacted))" (e.g., "redact")
   corrupted previously inserted markers, producing mangled output
   like "((((redacted))ed))".

Fix: use a bitmask to track redacted byte positions in the original
string. Consecutive masked regions collapse to a single "((redacted))"
marker in one final output pass.

Also use utf8.RuneCountInString for the single-char check to correctly
handle multi-byte characters.

Signed-off-by: Mathias Bogaert <mathias.bogaert@gmail.com>
